NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,729 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+			}
+		}
+	protected:
+		const int ThreadCount;
+		IMathEngine& MathEngine;
+		const CConstFloatHandle& First;
+		const int FirstHeight;
+		const int FirstWidth;
+		const CConstFloatHandle& Second;
+		const int SecondHeight;
+		const CFloatHandle& Result;
+	};
+
+	CThreadTaskMxMT task( owner.GetThreadPool()->Size(), mathEngine,
+		first, firstHeight, firstWidth, second, secondHeight, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CThreadTaskMxMT* )ptr )->CallRun( threadIndex );
+	} );
+}
+
+static void matrixDiagMultiply(
+	const CKMeansClustering& owner, IMathEngine& mathEngine,
+	const CConstFloatHandle& first, int firstSize,
+	const CConstFloatHandle& second, int secondWidth,
+	const CFloatHandle& result, int resultBufferSize )
+{
+	if( !isThreadTaskRelevant( firstSize * secondWidth ) ) {
+		mathEngine.MultiplyDiagMatrixByMatrix( first, firstSize,
+			second, secondWidth, result, resultBufferSize );
+		return;
+	}
+
+	NeoAssert( resultBufferSize >= firstSize * secondWidth );
+	struct CThreadTaskDiagMxM final {
+		CThreadTaskDiagMxM( IMathEngine& mathEngine, const CConstFloatHandle& first, int firstSize,
+				const CConstFloatHandle& second, int secondWidth, const CFloatHandle& result ) :
+			MathEngine( mathEngine ),
+			First( first ),
+			FirstSize( firstSize ),
+			Second( second ),
+			SecondWidth( secondWidth ),
+			Result( result )
+		{}
+
+		void CallRun( int /*threadIndex*/ )
+		{
+			for( int i = 0; i < FirstSize; ++i ) {
+				const CConstFloatHandle& multiplier = First + i;
+				MathEngine.VectorMultiply( Second + i * SecondWidth, Result + i * SecondWidth, SecondWidth, multiplier );
+			}
+		}
+	protected:
+		IMathEngine& MathEngine;
+		const CConstFloatHandle& First;
+		const int FirstSize;
+		const CConstFloatHandle& Second;
+		const int SecondWidth;
+		const CFloatHandle& Result;
+	};
+
+	CThreadTaskDiagMxM task( mathEngine, first, firstSize, second, secondWidth, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CThreadTaskDiagMxM* )ptr )->CallRun( threadIndex );
+	} );
+}
+
+static void vectorAddToMatrixRows(
+	const CKMeansClustering& owner, IMathEngine& mathEngine, int batchSize,
+	const CConstFloatHandle& matrix, const CFloatHandle& result,
+	int matrixHeight, int matrixWidth, const CConstFloatHandle& vector )
+{
+	NeoAssert( batchSize == 1 );
+
+	struct CVectorAddToMatrixRowsThreadTask final {
+		CVectorAddToMatrixRowsThreadTask( int threadCount, IMathEngine& mathEngine,
+				const CConstFloatHandle& matrix, const CFloatHandle& result,
+				int matrixHeight, int matrixWidth, const CConstFloatHandle& vector ) :
+			ThreadCount( threadCount ),
+			MathEngine( mathEngine ),
+			Matrix( matrix ),
+			MatrixHeight( matrixHeight ),
+			MatrixWidth( matrixWidth ),
+			Vector( vector ),
+			Result( result )
+		{}
+
+		void CallRun( int threadIndex )
```

Comment:
See comment about `CThreadTaskMxMT` and it's `Run/CallRun`

Reply:
fixed